lib/rules/no-duplicate-case.js

```diff
@@ -0,0 +1,57 @@
+/**
+ * @fileoverview Rule to disallow a duplicate case label.
+ * @author Dieter Oberkofler
+ * @copyright 2015 Dieter Oberkofler. All rights reserved.
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+
+    /**
+     * Get a hash value for the test
+     * @param {ASTNode} test The "test" node.
+     * @returns {string} A hash value for the test.
+     * @private
+     */
+    function getHash(test) {
+        if (test.type === "Literal") {
+            return test.type + typeof test.value + test.value;
+        } else if (test.type === "Identifier") {
+            return test.type + typeof test.name + test.name;
+        }
+    }
+
+    return {
+
+        "SwitchStatement": function(node) {
+
+            var caseCount = node.cases.length,
+                hashTable = {},
+                caseNode,
+                hashValue,
+                i;
+
+            for (i = 0; i < caseCount - 1; i++) {
```

Comment:
Sorry, that's not exactly what I mean. You don't really need loops at all, because estraverse is already has a built-in loop. What I had in mind is something like this code below (unfinished and untested, but hopefully gives you an idea).
```js
var switchStatement = [];

return {
    "SwitchStatement": function(node) { switchStatement.push({ node: node, caseLabels: []}); },
    "SwitchStatement:exit": function(node) { switchStatement.pop(); },
    "SwitchCase": function(node) {
        var currentSwitch = switchStatement[switchStatement.length - 1];
        if (node.test && currentSwitch.caseLabels[getHash(node.test)]) {
            context.report(node, "Duplicate case label.");
        } else if (node.test) {
            currentSwitch.caseLabels.push(getHash(node.test);
        }
    }
}
```